cli/azd/pkg/apphost/generate.go

```diff
@@ -899,6 +938,86 @@ func (b *infraGenerator) addContainer(
 	}
 }
 
+// buildContainer represents a container defined with a pre-build image or a build context.
+// container.v0 resources are used to define containers with pre-built images.
+//   - uses image field
+//
+// dockerfile.v0 resources are used to define containers with build context.
+//   - uses path and context fields
+//
+// container.v1 resources are used to define containers with either build context or pre-built images.
+//   - uses image field or build field
+func (b *infraGenerator) addBuildContainer(
+	name string,
+	r *Resource) error {
+	if r.Image != nil && r.Build != nil {
+		return fmt.Errorf("Resource '%s' cannot have both an image and a build", name)
+	}
+
+	b.requireCluster()
+	if len(r.Volumes) > 0 {
+		b.requireStorageVolume()
+	}
+
+	bc, err := buildContainerFromResource(r)
+	if err != nil {
+		return fmt.Errorf("container resource '%s': %w", name, err)
+	}
+	if bc.Build != nil {
+		b.requireContainerRegistry()
+	}
+	b.buildContainers[name] = *bc
+	return nil
+}
+
+func buildContainerFromResource(r *Resource) (*genBuildContainer, error) {
+	// common fields for all build containers
+	bc := &genBuildContainer{
+		Entrypoint: r.Entrypoint,
+		Args:       r.Args,
+		Env:        r.Env,
+		Bindings:   r.Bindings,
+		Volumes:    r.Volumes,
+	}
+
+	// container.v0 and container.v1+pre-build image
```

Comment:
Sort of wish that instead of doing `nil` tests on parts of the resource state to decide which version we were looking at we just did a switch over `r.Type` to determine which case we are in.  That would also let us add cases where you pass in some bogus resource type here accidently, via a `default` that returned an error?

Reply:
r.Type is always `container.v1` inside this function.  Depending on the state of container.v1 (build, context and image) we know what's the intention of the container.v1

Reply:
Got it - makes sense!